adr/0018-public-ids-for-qualifications.md

```diff
@@ -0,0 +1,71 @@
+# 17. Application Choice Updated At
+
+Date: 2021-01-28
+
+## Status
+
+In progress
+
+## Context
+
+The Vendor API exposes candidate qualifications and grades for systems to ingest.
+Each qualification in the system has an id, which vendor systems use to identify the qualification.
+It is therefore important that each qualification coming from the API has a unique id.
+We use the database id for a qualification object for this purpose.
+
+An issue arises when we have qualifications that are made up of multiple exams/grades.
+See [the ADR on structured grades](./0016-structured-gcse-grades.md) for background on the way gcse qualifications are stored.
+API consumers expect one qualification entity per exam/grade.
+However, because of the structured grades introduced above, some exams/grades share the same database id.
+Therefore, not all the ids for qualifications coming through the API are unique.
+
+## Decision
+
+We initially considered changing the data type of the id field in the API response to a string, to support ids of the form `123`, `123_1`, `123_2`.
+Vendors were not able to support that change, so the field had to remain an integer.
+
+The solution for this will be split into two: one for qualifications with a single grade in the API, and one for qualifications which have multiple entities in the API.
+The split is not as simple as structured vs unstructured grades since triple science is stored as a structure grade, but surfaced as a single GCSE in the API.
+The application_qualifications table will get a new public_id column to store the API id for a qualification.
+For qualifications with simple grades (only one exam/grade) (and triple science gcse), the public_id will just be set to the database id of the qualification initially via a data migration.
+For qualifications with structured grades (built from multiple exams/grades), the public_id column will be set to `nil`, and the public id will be set inside the structured_grades json, with one id per grade.
+A qualification with structured grades could look like this:
+```
+{
+    id: 123
+    subject: 'english',
+    public_id: nil,
+    grade: nil,
+    structured_grades: {
+      'English Language': { grade: 'E', public_id: '123' },
```

Comment:
Minor thing, the nested `public_id` type should probably match the global `public_id` type, i.e. they should probably both be integers (json supports ints).

Reply:
True, will update and ensure we do this!